File: .tmp_py4013/batch_12_dump/3341_find_minimum_time_to_reach_last_room_i.js
// LeetCode 3341 - Find Minimum Time to Reach Last Room I
// https://leetcode.com/problems/find-minimum-time-to-reach-last-room-i/

var minTimeToReach = function(moveTime) {
    const m = moveTime.length, n = moveTime[0].length;
    const dist = Array.from({length: m}, () => new Array(n).fill(1 << 30));
    const h = [[0, 0, 0]];
    dist[0][0] = 0;
    const dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    while (h.length) {
        h.sort((a, b) => a[0] - b[0]);
        const cur = h.shift();
        const t = cur[0], r = cur[1], c = cur[2];
        if (t !== dist[r][c]) continue;
        if (r === m - 1 && c === n - 1) return t;
        for (const d of dirs) {
            const nr = r + d[0], nc = c + d[1];
            if (nr < 0 || nc < 0 || nr >= m || nc >= n) continue;
            const start = Math.max(t, moveTime[nr][nc]);
            const nt = start + 1;
            if (nt < dist[nr][nc]) {
                dist[nr][nc] = nt;
                h.push([nt, nr, nc]);
            }
        }
    }
    return -1;
};

File: .tmp_rs4013/rust_check16/3341_find_minimum_time_to_reach_last_room_i.rs
struct Solution;
// LeetCode 3341 - Find Minimum Time to Reach Last Room I
// https://leetcode.com/problems/find-minimum-time-to-reach-last-room-i/

use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn min_time_to_reach(move_time: Vec<Vec<i32>>) -> i32 {
        let m = move_time.len();
        let n = move_time[0].len();
        let mut dist = vec![vec![1 << 30; n]; m];
        let mut h = BinaryHeap::new();
        h.push(Reverse((0, 0usize, 0usize)));
        dist[0][0] = 0;
        const DIRS: [[i32; 2]; 4] = [[0, 1], [1, 0], [0, -1], [-1, 0]];
        while let Some(Reverse((t, r, c))) = h.pop() {
            if t != dist[r][c] {
                continue;
            }
            if r == m - 1 && c == n - 1 {
                return t;
            }
            for d in &DIRS {
                let nr = r as i32 + d[0];
                let nc = c as i32 + d[1];
                if nr < 0 || nc < 0 || nr >= m as i32 || nc >= n as i32 {
                    continue;
                }
                let (nr, nc) = (nr as usize, nc as usize);
                let mut start = t;
                if move_time[nr][nc] > start {
                    start = move_time[nr][nc];
                }
                let nt = start + 1;
                if nt < dist[nr][nc] {
                    dist[nr][nc] = nt;
                    h.push(Reverse((nt, nr, nc)));
                }
            }
        }
        -1
    }
}

fn main() {}
